cmd/signimage/signimage.go

```diff
@@ -51,38 +51,88 @@ func signFile(filename string, publickey string, privatekey string) error {
 	return nil
 }
 
-func getAuthFromFile(configfile string, repo string) (authn.Authenticator, error) {
+// allSecretFiles  is a comma seperated list of files and directories to search for creds for repo
+// so we need to search through them for the right one.
+// if we get an error we just log it and move on to the next file/dir i
+// and if we find nothing try anonymous auth
+// we're tolerant in our inputs and precise in our inputs. Its the UNIX way!
+func getAuthFromFileList(allSecretFiles string, repo string) authn.Authenticator {
+	var auth authn.Authenticator
+	for _, configFile := range strings.Split(allSecretFiles, ",") {
+		fileInfo, err := os.Stat(configFile)
+		if err != nil {
+			logger.Info("unable to Stat ", "filename", configFile, "error=", err)
+			continue
+		}
+		if fileInfo.IsDir() {
+			files, err := os.ReadDir(configFile)
+			if err != nil {
+				logger.Info("unable to read directory", "dirname", configFile, "error=", err)
+			}
 
-	if configfile == "" {
-		logger.Info("no pull secret defined, default to Anonymous")
-		return authn.Anonymous, nil
+			for _, file := range files {
+				if file.Name()[0:2] == ".." || file.Name() == "." {
+					continue
+				}
+				auth, err = getAuthFromSingleFile(configFile+"/"+file.Name(), repo)
+				if err != nil {
+					logger.Info("unable to read file from dir", "dirname", configFile, "error=", err)
+					continue
+				}
+			}
+		} else {
+			auth, err = getAuthFromSingleFile(configFile, repo)
+			if err != nil {
+				logger.Info("unable to read file", "filename", configFile, "error=", err)
+			}
+		}
+		if auth != nil {
+			logger.Info("Successfully found secret", "repo", repo, "filename", configFile)
+			return auth
+		}
 	}
+	logger.Info("no secret found, default to Anonymous")
+	return authn.Anonymous
+}
 
-	f, err := os.Open(configfile)
+func getAuthFromSingleFile(secretFile string, repo string) (authn.Authenticator, error) {
+	jsonFile, err := os.Open(secretFile)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to open secret file %s: %v", secretFile, err)
 	}
-	defer f.Close()
-	cf, err := config.LoadFromReader(f)
+	// defer the closing of our jsonFile so that we can parse it later o
+	defer jsonFile.Close()
+
+	byteValue, err := io.ReadAll(jsonFile)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error reading secret file %s: %v", secretFile, err)
 	}
 
-	var cfg dockertypes.AuthConfig
+	var result map[string]json.RawMessage
+	var allAuthConfigs map[string]authn.AuthConfig
 
-	cfg, err = cf.GetAuthConfig(repo)
+	err = json.Unmarshal(byteValue, &result)
```

Comment:
Wondering if it would make sense to instead feed `jsonFile` to a [`json.Decoder`](https://pkg.go.dev/encoding/json#Decoder). We potentially need to read it again, so we would need to `seek(0)`, but we still avoid reading the entire file in memory.

Reply:
Would decoder not read the file into memory?  Or am I missing your point?

I'm opening and closing the file, then (potentially) opening and closing it again, which is slower then opening it once and seeking, but performance isn't really an issue here, (the whole signing job runs in a few seconds). I'm also quitting the search when I find the first authenticator that matches the repo whcih speeds things up but means we only loop through as many files as we need to, so on the second iteration we can't be sure if a file has been opened previously (so needed seek(0) ) and what hadn't (so needs open() ). 
It sounds more complicated and I'm not sure I understand the benefit.

Also the max size of a k8s secret is 1mb so thats all holding the file in memory could potentially need.

Reply:
> Would decoder not read the file into memory?

The marshalled contents would, but not the bytes themselves - now, I believe we have both.
However I agree that the suggestion is probably premature optimization (if it optimizes anything at all) and doesn't deserve more time now.